### ISSUE TITLE
Reproduction: CUDA Port with BPTT

### ISSUE THREAD

**Issue by LifeSupport0** [jrz97619761/test-model-thing#24]
Ported the model to native CUDA on Linux (because I can, not because I hate MLX).
Used BPTT for better 
Run on an NVidia RTX 3080 (10GB) card, averaging ~20 chunks/s (64B/chunk)
Note: a prior, more faithful version of the training program was latency-bound and capped out at 8 chunks/s

The entire training set, model, and optimizer fit 100% resident on the GPU, `nvidia-smi` reports 61% compute utilization over the training run.

Approximate training time: 928,000 steps (checkpoint per 1k steps)

Clock time: \~11 hours

Corpus Size: \~3GB (majority of simplewiki-20260801)

Epochs: \~1

Apparent training loss oscillates between 1.0 and 2.1, for particular batches.
The model has converged, and was fine-tuning on the checkpoints when stopped.
Oscillation is driven by CE, not model thrashing.

CoLA benchmark run
```
$ python ./benchmark.py ./tmt.pt 1 0.5 
Starting benchmark.

Epoch 1 / 1
[500 / 4274] train: T+ 293, T- 14, F+ 158, F- 36 (-4.4398)
[1000 / 4274] train: T+ 570, T- 52, F+ 303, F- 76 (4.1304)
[1500 / 4274] train: T+ 864, T- 83, F+ 442, F- 112 (6.1477)
[2000 / 4274] train: T+ 1112, T- 125, F+ 593, F- 171 (5.5140)
[2500 / 4274] train: T+ 1474, T- 126, F+ 725, F- 176 (6.0191)
[3000 / 4274] train: T+ 1745, T- 166, F+ 872, F- 218 (6.9586)
[3500 / 4274] train: T+ 2085, T- 175, F+ 1015, F- 226 (7.3277)
[4000 / 4274] train: T+ 2462, T- 179, F+ 1126, F- 234 (7.7612)
[4274 / 4274] train: T+ 2645, T- 180, F+ 1210, F- 240 (7.2878)
[500 / 4275] held: T+ 331, T- 1, F+ 168, F- 1 (2.1783)
[1000 / 4275] held: T+ 705, T- 1, F+ 294, F- 1 (2.0148)
[1500 / 4275] held: T+ 1075, T- 3, F+ 421, F- 2 (4.0774)
[2000 / 4275] held: T+ 1462, T- 3, F+ 533, F- 3 (2.8747)
[2500 / 4275] held: T+ 1846, T- 3, F+ 649, F- 3 (2.6731)
[3000 / 4275] held: T+ 2230, T- 4, F+ 764, F- 3 (3.4961)
[3500 / 4275] held: T+ 2578, T- 6, F+ 909, F- 8 (2.4115)
[4000 / 4275] held: T+ 2894, T- 28, F+ 1024, F- 55 (2.4603)
[4275 / 4275] held: T+ 3062, T- 44, F+ 1094, F- 76 (3.8654)
```

**Comment by SimonVitzethum** [comment]
Nice — good to see another native CUDA port. I've been down the same road (mine's CUDA with the kernels written in Zig and compiled through LLVM to PTX), so a few numbers that might save you some time.

Throughput, for calibration. I get 132 kB/s at dim=512 layers=16 experts=8 topk=2 batch=64 seqlen=128, on an RTX 5070 Laptop that's power-capped at 50–55 W of its 115 W. Your 3 GB in 11 h works out to ~76 kB/s. Different model sizes, so not apples to apples — what's your dim/layers/experts? Also, 928k steps over 3 GB is 3.2 kB/step, which doesn't square with 64 B/chunk unless a chunk carries a batch dimension of ~50. Worth stating, because the two numbers get compared.

I'd distrust the 61 % compute utilization. nvidia-smi's utilization.gpu is the fraction of time at least one kernel was resident — not occupancy, not throughput. I measured 99–100 % on that metric while the GPU was actually idle for over half the step. The numbers that mean something are bytes moved against a measured streaming-copy ceiling, and FLOPs against a measured large-GEMM ceiling, both benchmarked on the same card. Mine came out at 82 % of the bandwidth ceiling while the kernels run, and 15 % of the matmul ceiling. The arithmetic intensity is ~63 FLOP/byte on a machine that needs ~138 to balance the two, so this workload is bandwidth-bound by construction — no amount of kernel tuning moves the compute number, and chasing it is wasted effort.

On the earlier latency-bound version capping at 8 chunks/s — same experience, and the causes were findable:

- A cudaDeviceSynchronize after every driver-API launch. Removing it: +19 %.
- A fixed per-step cost that doesn't scale with the batch. I found it by fitting wall time against batch size: 8× the batch gave only 2× the throughput, which isolates ~8.3 ms of per-step overhead. At batch=16 that's 35 % of the step; at batch=64 it's 12 %. It's ~360 kernel launches at ~2 µs plus cuBLAS calls at ~10 µs each.
- Worth measuring before optimizing anything: I spent a while attacking the expert matmuls because my cost model said they dominated, then forced the batched path and the throughput moved 0.8 %. The model was wrong, not the code.

One epoch on 3 GB is the right call, and there's a measurement behind that. At 95 MB against ~34M parameters — 2.4 bytes per parameter — I got 1.63 BPB held out at 14 epochs and 1.82 at 24. More epochs made it actively worse. If you're at ~1 epoch on 3 GB you're in the regime where the data is doing the work; I've just moved to enwik9 for the same reason.

I'd hold off on "converged", though. A batch loss oscillating 1.0–2.1 on simplewiki is mostly per-batch CE variance over heterogeneous articles — you're right that it isn't thrashing, but it also doesn't show convergence. The curve that does is held-out BPB on a fixed slice over time. For a yardstick at byte level: Qwen3-1.7B scores 0.853 BPB on a held-out slice of enwik8, so there's a lot of headroom below 1.0.

One thing that may bite you with BPTT specifically: if your longest half-life exceeds the BPTT window, those channels get systematically truncated gradients. Mine ran half_max=65536 at seqlen=128 — a factor of 512. The state carries across windows but the gradient doesn't, so the slow channels hold context they were never trained to fill. Worth checking what your ratio is.
